src/atomic_c.c

```diff
@@ -118,306 +118,139 @@
-
-#undef SHMEM_CAPI_ATOMIC_SET
+#define SHMEM_DEF_CSWAP(STYPE,TYPE,ITYPE) \
+    TYPE shmem_##STYPE##_cswap(TYPE *target, TYPE cond, TYPE value,     \
+                               int pe)                                  \
+    {                                                                   \
+        TYPE newval;                                                    \
+        SHMEM_ERR_CHECK_INITIALIZED();                                  \
+        SHMEM_ERR_CHECK_PE(pe);                                         \
+        SHMEM_ERR_CHECK_SYMMETRIC(target, sizeof(TYPE));                \
+        shmem_internal_cswap(target, &value, &newval, &cond,            \
+                             sizeof(TYPE), pe, ITYPE);                  \
+        shmem_internal_get_wait();                                      \
+        return newval;                                                  \
+    }
+
+SHMEM_BIND_C_AMO(SHMEM_DEF_CSWAP)
+
+#define SHMEM_DEF_INC(STYPE,TYPE,ITYPE)                                 \
+    void shmem_##STYPE##_inc(TYPE *target, int pe)                      \
+    {                                                                   \
+        TYPE tmp = 1;                                                   \
+        SHMEM_ERR_CHECK_INITIALIZED();                                  \
+        SHMEM_ERR_CHECK_PE(pe);                                         \
+        SHMEM_ERR_CHECK_SYMMETRIC(target, sizeof(TYPE));                \
+        shmem_internal_atomic_small(target, &tmp, sizeof(TYPE), pe,     \
+                                    SHM_INTERNAL_SUM, ITYPE);           \
+    }
+
+SHMEM_BIND_C_AMO(SHMEM_DEF_INC)
+
+#define SHMEM_DEF_FINC(STYPE,TYPE,ITYPE)                                \
+    TYPE shmem_##STYPE##_finc(TYPE *target, int pe)                     \
+    {                                                                   \
+        TYPE oldval, tmp = 1;                                           \
+        SHMEM_ERR_CHECK_INITIALIZED();                                  \
+        SHMEM_ERR_CHECK_PE(pe);                                         \
+        SHMEM_ERR_CHECK_SYMMETRIC(target, sizeof(TYPE));                \
+        shmem_internal_fetch_atomic(target, &tmp, &oldval, sizeof(TYPE),\
+                                    pe, SHM_INTERNAL_SUM, ITYPE);       \
+        shmem_internal_get_wait();                                      \
+        return oldval;                                                  \
+    }
+
+SHMEM_BIND_C_AMO(SHMEM_DEF_FINC)
+
+#define SHMEM_DEF_ADD(STYPE,TYPE,ITYPE)                                 \
+    void shmem_##STYPE##_add(TYPE *target, TYPE value, int pe)          \
+    {                                                                   \
+        SHMEM_ERR_CHECK_INITIALIZED();                                  \
+        SHMEM_ERR_CHECK_PE(pe);                                         \
+        SHMEM_ERR_CHECK_SYMMETRIC(target, sizeof(TYPE));                \
```

Comment:
you could further refactor...you probably don't want to at this point, by creating an arg_check function to put all these checks into =) save of 2 lines per repeat